File: accidentes/AMarotoChivite-Accidentes/AMarotoChivite-Accidentes-Java/src/main/java/Ficheros/Accidentes/controllers/ControllerQueries.java
package Ficheros.Accidentes.controllers;

import Ficheros.Accidentes.models.dto.AccidenteDto;
import Ficheros.Accidentes.utils.Triple;
import ch.qos.logback.classic.Logger;
import org.slf4j.LoggerFactory;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.stream.Collectors;

import static Ficheros.Accidentes.utils.ParseLocalDate.parseFecha;
import static java.util.stream.Collectors.groupingBy;

public class ControllerQueries {
    private static final Logger logger = (Logger) LoggerFactory.getLogger(ControllerQueries.class);
    private final List<AccidenteDto> listOfStorageReadFile;
    private final Map<String, List<AccidenteDto>> mapByDistrito;

    public ControllerQueries(List<AccidenteDto> listOfStorageReadFile) {
        this.listOfStorageReadFile = listOfStorageReadFile;

        this.mapByDistrito = listOfStorageReadFile.stream()
                .collect(groupingBy(it -> it.getDistrito()));
                /* OTRA FORMA (reflexiva): .collect(Collectors.groupingBy(AccidenteDto::getDistrito)); */
        logger.debug("ControllerQueries: CARGANDO CONSULTAS");
    }

    public List<AccidenteDto> getAccidentesOnlyAlcoholOrDrugs() {
        return listOfStorageReadFile.stream()
                .filter(accidente -> !"NULL".equals(accidente.getPositivaDroga()) || !"N".equals(accidente.getPositivaAlcohol()))
                .collect(Collectors.toList());
    }

    public int numAccidentesOnlyAlcoholAndDrugs() {
        return (int) listOfStorageReadFile.stream()
                .filter(a -> !"NULL".equals(a.getPositivaDroga()) && !"N".equals(a.getPositivaAlcohol()))
                .count();
    }

    public Map<String, List<AccidenteDto>> accidentesBySexo() {
        return listOfStorageReadFile.stream()
                .collect(Collectors.groupingBy(AccidenteDto::getSexo));
    }

    public Map<String, List<AccidenteDto>> accidentesByMonths() {
        return listOfStorageReadFile.stream()
                .collect(Collectors.groupingBy(dto -> dto.getFecha().substring(3, 5)));
    }

    public Map<String, List<AccidenteDto>> accidentesByVehiculos() {
        return listOfStorageReadFile.stream()
                .collect(Collectors.groupingBy(AccidenteDto::getTipoVehiculo));
    }

    // EN LEGANÉS
    public List<AccidenteDto> accidentesByLocalizacion(String localizacion) {
        return listOfStorageReadFile.stream()
                .filter(dto -> dto.getLocalizacion().contains(localizacion.toUpperCase(Locale.getDefault())))
                .collect(Collectors.toList());
    }

    public Map<String, Integer> numAccidentesByDistrito() {
        return mapByDistrito.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().size()));
    }

    public List<AccidenteDto> estadisticasByDistrito() {
        throw new UnsupportedOperationException("Not implemented yet");
    }

    public Map<String, List<AccidenteDto>> accidentesByDistritoOrderDesc(int numeroAccidentersParaVisualizar) {
        throw new UnsupportedOperationException("Not implemented yet");
/*
        return mapByDistrito.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().stream()
                        .sorted(Comparator.comparing(AccidenteDto::getFecha).reversed())
                        .limit(numeroAccidentersParaVisualizar)
                        .collect(Collectors.toList())))
                .entrySet().stream()
                .sorted(Map.Entry.comparingByKey(Comparator.reverseOrder()))
                .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, (v1, v2) -> v1, LinkedHashMap::new));*/
    }

    public List<AccidenteDto> accidentesEnFinDeSemanaAndNoche() {
        return listOfStorageReadFile.stream()
                .filter(accidente -> {
                    LocalDate fecha = parseFecha(accidente.getFecha());
                    boolean esFinDeSemana = fecha.getDayOfWeek() == DayOfWeek.SATURDAY || fecha.getDayOfWeek() == DayOfWeek.SUNDAY;
                    DateTimeFormatter flexibleFormatter = DateTimeFormatter.ofPattern("[H:]mm[:ss][ a]");
                    LocalTime hora = LocalTime.parse(accidente.getHora(), flexibleFormatter);
                    return esFinDeSemana && (hora.isAfter(LocalTime.of(19, 59)) || hora.isBefore(LocalTime.of(6, 0)));
                })
                .collect(Collectors.toList());
    }

    public List<AccidenteDto> accidentesEnFinDeSemanaAndNocheAndPositiveAlcohol() {
        List<AccidenteDto> listFinDeSemanaNoche = accidentesEnFinDeSemanaAndNoche();

        return listFinDeSemanaNoche.stream()
                .filter(accidente -> !accidente.getPositivaDroga().equals("NULL") && !accidente.getPositivaAlcohol().equals("N"))
                .collect(Collectors.toList());
    }

    public List<AccidenteDto> accidentesWhereDead() {
        return listOfStorageReadFile.stream()
                .filter(accidente -> accidente.getLesividad().contains("Fallecido"))
                .collect(Collectors.toList());
    }

    public Triple<Boolean, String, String> isDistritoMoreRiskySameAsDistritoMoreRiskyInWeekend() {
        Map<String, Integer> numAccidentesByDistrito = numAccidentesByDistrito();
        String distritoMoreAccidentes = Collections.max(numAccidentesByDistrito.entrySet(),
                Comparator.comparingInt(Map.Entry::getValue)).getKey();

        String distritoMoreAccidentesWeekend = mapByDistrito.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().stream()
                        .filter(accidente -> parseFecha(accidente.getFecha()).getDayOfWeek() == DayOfWeek.SATURDAY
                                || parseFecha(accidente.getFecha()).getDayOfWeek() == DayOfWeek.SUNDAY)
                        .collect(Collectors.toList())))
                .entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, e -> e.getValue().size()))
                .entrySet().stream()
                .max(Map.Entry.comparingByValue())
                .get()
                .getKey();

        if (distritoMoreAccidentes.equals(distritoMoreAccidentesWeekend)) {
            return new Triple<>(true, distritoMoreAccidentes, distritoMoreAccidentesWeekend);
        } else {
            return new Triple<>(false, distritoMoreAccidentes, distritoMoreAccidentesWeekend);
        }
    }

    public int numAccidentesWherePositiveAlcholoOrDrugsAndDeads() {
        return (int) listOfStorageReadFile.stream()
                .filter(accidente -> (!"NULL".equals(accidente.getPositivaDroga()) || !"N".equals(accidente.getPositivaAlcohol())) && accidente.getLesividad().contains("Fallecido"))
                .count();
    }

    public List<AccidenteDto> numAccidentesWhereHitPeople() {
        return listOfStorageReadFile.stream()
                .filter(accidente -> accidente.getTipoAccidente().contains("persona"))
                .collect(Collectors.toList());
    }

    public Map<String, List<AccidenteDto>> accidentesByMeteorologia() {
        return listOfStorageReadFile.stream()
                .collect(Collectors.groupingBy(AccidenteDto::getEstadoMeteorologico));
    }

    public List<AccidenteDto> accidentesWhereHitAnimal() {
        return listOfStorageReadFile.stream()
                .filter(accidente -> accidente.getTipoAccidente().contains("animal"))
                .collect(Collectors.toList());
    }
}

File: accidentes/AMarotoChivite-Accidentes/AMarotoChivite-Accidentes-Kotlin/src/main/kotlin/Ficheros/Accidentes/controllers/ControllerQueries.kt
package Ficheros.Accidentes.controllers

import Ficheros.Accidentes.models.dto.AccidenteDTO
import Ficheros.Accidentes.utils.parseFecha
import mu.KotlinLogging
import java.time.DayOfWeek
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.util.*

class ControllerQueries(private val listOfStorageReadFile: List<AccidenteDTO>) {

    private val logger = KotlinLogging.logger {}.debug { "ControllerQueries: CARGANDO CONSULTAS)" }

    private val mapByDistrito = listOfStorageReadFile.groupBy { it.distrito }

    fun getAccidentesOnlyAlcoholOrDrugs(): List<AccidenteDTO> {
        return listOfStorageReadFile.filter { it.positivaDroga != "NULL" || it.positivaAlcohol != "N" }
    }

    fun numAccidentesOnlyAlcoholAndDrugs(): Int {
        return listOfStorageReadFile.count { it.positivaDroga != "NULL" && it.positivaAlcohol != "N" }
    }

    fun accidentesBySexo(): Map<String, List<AccidenteDTO>> {
        return listOfStorageReadFile.groupBy { it.sexo }
    }

    fun accidentesByMonths(): Map<String, List<AccidenteDTO>> {
        return listOfStorageReadFile.groupBy { it.fecha.substring(3, 5) }
    }

    fun accidentesByVehiculos(): Map<String, List<AccidenteDTO>> {
        return listOfStorageReadFile.groupBy { it.tipoVehiculo }
    }

    // EN LEGANÉS
    fun accidentesByLocalizacion(localizacion: String): List<AccidenteDTO> {
        return listOfStorageReadFile.filter { it.localizacion.contains(localizacion.uppercase(Locale.getDefault())) }
    }

    fun numAccidentesByDistrito(): Map<String, Int> {
        return mapByDistrito.mapValues { (_, list) -> list.size }
    }

    fun estadisticasByDistrito(): List<AccidenteDTO> {
        TODO()
    }

    fun accidentesByDistritoOrderDesc(numeroAccidentersParaVisualizar: Int): Map<String, List<AccidenteDTO>> {
        return mapByDistrito.mapValues { (_, group) ->
            group.take(numeroAccidentersParaVisualizar)
        }.toSortedMap(Comparator.reverseOrder())
    }

    // (a partir de las 8 de la tarde hasta las 6 de la mañana)
    fun accidentesEnFinDeSemanaAndNoche(): List<AccidenteDTO> {
        return listOfStorageReadFile.filter { accidente ->
            val fecha = parseFecha(accidente.fecha)
            val esFinDeSemana = fecha.dayOfWeek == DayOfWeek.SATURDAY || fecha.dayOfWeek == DayOfWeek.SUNDAY
            // No lo puede parsear! Por ello de manera opcional acepta "h:mm" como "HH:mm"
            val flexibleFormatter = DateTimeFormatter.ofPattern("[H:]mm[:ss][ a]")
            val hora = LocalTime.parse(accidente.hora, flexibleFormatter)

            esFinDeSemana && (hora >= LocalTime.of(20, 0) || hora < LocalTime.of(6, 0))
        }
    }

    fun accidentesEnFinDeSemanaAndNocheAndPositiveAlcohol(): List<AccidenteDTO> {
        val listFinDeSemanaNoche = accidentesEnFinDeSemanaAndNoche()

        return listFinDeSemanaNoche.filter {
            it.positivaDroga != "NULL" && it.positivaAlcohol != "N"
        }
    }

    fun accidentesWhereDead(): List<AccidenteDTO> {
        return listOfStorageReadFile.filter { it.lesividad.contains("Fallecido") }
    }

    // Comprobar si el distrito con más accidentes coincide con el distrito donde hay más accidentes en fin de semana.
    fun isDistritoMoreRiskySameAsDistritoMoreRiskyInWeekend(): Triple<Boolean, String, String> {
        val numAccidentesByDistrito = numAccidentesByDistrito()
        val distritoMoreAccidentes = numAccidentesByDistrito.maxByOrNull { it.value }!!.key

        val distritoMoreAccidentesWeekend = mapByDistrito.mapValues {
            it.value.filter { accidente ->
                parseFecha(accidente.fecha).dayOfWeek == DayOfWeek.SATURDAY
                        || parseFecha(accidente.fecha).dayOfWeek == DayOfWeek.SUNDAY
            }
        }.mapValues { (_, list) -> list.size }
            .maxByOrNull { it.value }!!.key

        if (distritoMoreAccidentes == distritoMoreAccidentesWeekend) {
            return Triple(true, distritoMoreAccidentes, distritoMoreAccidentesWeekend)
        } else {
            return Triple(false, distritoMoreAccidentes, distritoMoreAccidentesWeekend)
        }
    }

    fun numAccidentesWherePositiveAlcholoOrDrugsAndDeads(): Int {
        return listOfStorageReadFile.count {
            (it.positivaDroga != "NULL" || it.positivaAlcohol != "N") && it.lesividad.contains("Fallecido")
        }
    }

    fun numAccidentesWhereHitPeople(): List<AccidenteDTO> {
        return listOfStorageReadFile.filter { it.tipoAccidente.contains("persona") }
    }

    fun accidentesByMeteorologia(): Map<String, List<AccidenteDTO>> {
        return listOfStorageReadFile.groupBy { it.estadoMeteorologico }
    }

    fun accidentesWhereHitAnimal(): List<AccidenteDTO> {
        return listOfStorageReadFile.filter { it.tipoAccidente.contains("animal") }
    }

}